kismet: fix cross compilation

pete3n opened the pull request:
Changes to allow cross-compile builds with pkgsCross 
See: https://gist.github.com/pete3n/4d997ee43277d31e2d7ee2a987018f5f

## Description of changes

- Added pre-configure export of pkg-config path to prevent breaking in configure script during cross-builds
- Added config flags for protoc and protoc-c to prevent missing dependencies during cross-builds
- Set withPython default to false to prevent cross-build breaking issues I have not solved yet

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-linux -> aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x  Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pete3n committed:
kismet: fix cross compilation

Update default.nix

kismet: fix cross compilation

NickCao commented on pkgs/applications/networking/sniffers/kismet/default.nix:
```diff
@@ -39,6 +39,10 @@ stdenv.mkDerivation rec {
   '';
 
+  preConfigure = ''
+    export PATH=${pkg-config}/bin/:$PATH
```
This adds the `hostPlatform` `pkgs-config` to `PATH`, which (is most cases) is not runnable on `buildPlatform`.

pete3n replied:
I just want to clarify: Are you saying this breaks most builds or that this fixes most builds? Native builds are currently fine for the existing package on both aarch64 and x86_64 (probably most builds). Cross builds are currently broken (specifically x86_64 -> aarch64, but probably all cross-builds), which is what my pull request addresses.

NickCao replied:
This would not break native build, but still won't fix cross build. If it works for you, could be binfmt_misc interfering, please disable that and try again.

NickCao replied:
A proper fix is to replace all hardcoded references to `pkg-config` with `$PKG_CONFIG`, which points to the pkg-config binary (something like `aarch64-unknown-linux-gnu-pkg-config`) of the correct host/target platforms.

pete3n replied:
> This would not break native build, but still won't fix cross build. If it works for you, could be binfmt_misc interfering, please disable that and try again.

Aw, you are right. Good catch. I've been emulating pkg-config and the protoc compilers without realizing it :-|. This does not work without qemu-user-static package on Debian based machines or  boot.binfmt.emulatedSystems = [ "aarch64-linux" ] on NixOS. Still, this is the only way I have been able to semi-cross-compile on Non-NixOS machines. 

I guess if this isn't really suitable for a merge, I can create an issue, and then submit this gist: https://gist.github.com/pete3n/4d997ee43277d31e2d7ee2a987018f5f  as a solution to the issue? I just don't want other people to struggle the way I have to build this with pkgsCross.

pete3n replied:
Sumitted an issue: https://github.com/NixOS/nixpkgs/issues/276165

That can be closed with either the provided gits or possibly a revised version of this pull request. Obviously at your discretion. I just want to provide some workarounds to my cross-build issue.

NickCao commented:
Try this:

```diff
--- a/pkgs/applications/networking/sniffers/kismet/default.nix
+++ b/pkgs/applications/networking/sniffers/kismet/default.nix
@@ -1,5 +1,6 @@
 { lib
 , stdenv
+, autoreconfHook
 , binutils
 , elfutils
 , fetchurl
@@ -20,7 +21,7 @@
 , python3
 , sqlite
 , withNetworkManager ? false
-, withPython ? true
+, withPython ? stdenv.buildPlatform.canExecute stdenv.hostPlatform
 , withSensors ? false
 , zlib
 }:
@@ -37,6 +38,8 @@ stdenv.mkDerivation rec {
   postPatch = ''
     substituteInPlace Makefile.in \
       --replace "-m 4550" ""
+    substituteInPlace configure.ac \
+      --replace "pkg-config" "$PKG_CONFIG"
   '';
 
   postConfigure = ''
@@ -47,8 +50,13 @@ stdenv.mkDerivation rec {
         -i Makefile
   '';
 
+  strictDeps = true;
+
   nativeBuildInputs = [
+    autoreconfHook
     pkg-config
+    protobuf
+    protobufc
   ] ++ lib.optionals withPython [
     python3
   ];
```

pete3n commented:
Nick, 
Thanks very much for  this. Here are my results:
```
, withPython ? stdenv.buildPlatform.canExecute stdenv.hostPlatform
```
This seems to have inverted the previous results. Now cross-compile builds run with pkgsCross build but native builds with pkgs throw the Python dependency error previously seen with the cross-builds.

## Builds 
### x86_64-linux NixOS 23.11 (no binfmt emulation)
- pkgsCross.aarch64-multiplatform.kismet [x]
- pkgs.kismet [] -- fails with Python dependency
- pkgs.kismet - withPython false [x]

### x86_64-linux Debian 11 (no qemu-user-static)
- pkgsCross.aarch64-multiplatform.kismet [x]
- pkgs.kismet [] -- fails with Python dependency
- pkgs.kismet - withPython false [x]

### aarch64-linux Ubuntu 22.04
- pkgs.kismet [] -- fails with Python dependency
- pkgs.kismet - withPython false [x]

NickCao commented:
> This seems to have inverted the previous results. 

Oh that was a typo, try inverting it again.

NickCao commented:
> Now cross-compile builds run with pkgsCross succeed but native builds with pkgs throw the Python dependency error previously seen with the cross-builds.

This could be the `strictDeps`.

pete3n commented:
> > Now cross-compile builds run with pkgsCross succeed but native builds with pkgs throw the Python dependency error previously seen with the cross-builds.
> 
> This could be the `strictDeps`. Try moving everything python to `nativeBuildInputs`

That was it. I moved the conditional Python dependencies into the nativeBuildInputs and now all build scenarios are completing. I updated the PR with these changes. Thanks for your help. I'll be happy to see this get merged up.

hughobrien commented:
Result of `nixpkgs-review pr 276038` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kismet</li>
  </ul>
</details>

NickCao commented on pkgs/applications/networking/sniffers/kismet/default.nix:
```diff
@@ -47,10 +50,22 @@ stdenv.mkDerivation rec {
+    protobuf
+    protobufc
   ] ++ lib.optionals withPython [
     python3
```
```suggestion
```

This is redundant.

NickCao commented:
Also please squash the commits and fix the commit message, it should look like `kismet: fix cross compilation`

NickCao reviewed (approved):
diff LGTM, future works could be re-enabling python support for cross compilation. Waiting for ofborg.
